leakcanary-android-release/src/main/java/leakcanary/ConditionalHeapAnalyzer.kt

```diff
@@ -0,0 +1,114 @@
+package leakcanary
+
+import android.app.Application
+import leakcanary.HeapAnalysisCondition.Result.StopAnalysis
+import leakcanary.HeapAnalysisCondition.Trigger
+import leakcanary.internal.ReleaseHeapAnalyzer
+import leakcanary.internal.startBackgroundHandlerThread
+import leakcanary.internal.uiHandler
+import shark.HeapAnalysis
+import shark.SharkLog
+
+/**
+ * Performs automatic heap dump & analysis when conditions are met.
+ */
+class ConditionalHeapAnalyzer(
+  application: Application,
+  private val conditions: List<HeapAnalysisCondition> = defaultConditions(application),
+  config: HeapAnalysisConfig = HeapAnalysisConfig(),
+  private val listener: Listener
+) {
+
+  interface Listener {
+    /**
+     * Called on the main thread.
+     */
+    fun onConditionsUpdate(result: Result)
+  }
+
+  sealed class Result {
+    class HeapAnalysisResult(val heapAnalysis: HeapAnalysis) : Result() {
+      override fun toString(): String {
+        return heapAnalysis.toString()
+      }
+    }
+
+    data class ConditionsNotMet(val reason: String) : Result()
+  }
+
+  private val heapAnalyzer = ReleaseHeapAnalyzer(application, config) { analysis ->
+    uiHandler.post {
+      listener.onConditionsUpdate(Result.HeapAnalysisResult(analysis))
+    }
+    conditionTrigger.conditionChanged("analysis done")
+  }
+
+  @Volatile
+  private var enabled = false
+
+  private val conditionTrigger: Trigger = object : Trigger {
+    override fun conditionChanged(reason: String) {
+      if (!enabled) {
+        return
+      }
+      SharkLog.d { "Rescheduling heap analysis condition check because $reason" }
+      backgroundHandler.removeCallbacks(checkConditions)
+      backgroundHandler.post(checkConditions)
+    }
+  }
+
+  private val checkConditions = Runnable {
+    if (!enabled) {
+      return@Runnable
+    }
+    conditions.forEach { condition ->
+      condition.lateInitTrigger = conditionTrigger
+      val result = condition.evaluate()
+      if (result is StopAnalysis) {
+        heapAnalyzer.stop()
+        uiHandler.post {
+          listener.onConditionsUpdate(Result.ConditionsNotMet(result.reason))
+        }
+        return@Runnable
+      }
+      if (!enabled) {
+        return@Runnable
+      }
```

Comment:
Shouldn't this go before `val result = condition.evaluate()` ?

Reply:
There's already a check before the first condition. This adds a check in between every condition check and also ensures there's a check after the last one.

If I moved this before `val result = condition.evaluate()` I'd have to move the very first check to the very end of the method. same same?